govcd/user.go

```diff
@@ -0,0 +1,505 @@
+	OrgUserRoleVappUser                  = "vApp User"
+	OrgUserRoleConsoleAccessOnly         = "Console Access Only"
+	OrgUserRoleDeferToIdentityProvider   = "Defer to Identity Provider"
+
+	// Allowed values for provider types
+	OrgUserProviderIntegrated = "INTEGRATED" // The user is created locally or imported from LDAP
+	OrgUserProviderSAML       = "SAML"       // The user is imported from a SAML identity provider.
+	OrgUserProviderOAUTH      = "OAUTH"      // The user is imported from an OAUTH identity provider
+)
+
+// Used to check the validity of provider type on creation
+var OrgUserProviderTypes = []string{
+	OrgUserProviderIntegrated,
+	OrgUserProviderSAML,
+	OrgUserProviderOAUTH,
+}
+
+// NewUser creates an empty user
+func NewUser(cli *Client, org *AdminOrg) *OrgUser {
+	return &OrgUser{
+		User:     new(types.User),
+		client:   cli,
+		AdminOrg: org,
+	}
+}
+
+// Private function to perform the user search by either name or ID
+// It will only look for one of the two fields, depending on the value of byName
+func (adminOrg *AdminOrg) fetchUserByNameOrId(identifier string, byName, refresh bool) (*OrgUser, error) {
+	if refresh {
+		err := adminOrg.Refresh()
+		if err != nil {
+			return nil, err
+		}
+	}
+	var userHREF string
+
+	for _, orgUser := range adminOrg.AdminOrg.Users.User {
+		if !byName && orgUser.ID == identifier {
+			userHREF = orgUser.HREF
+			break
+		}
+		if byName && orgUser.Name == identifier {
+			userHREF = orgUser.HREF
+			break
+		}
+	}
+
+	if userHREF != "" {
+		orgUser := NewUser(adminOrg.client, adminOrg)
+
+		_, err := adminOrg.client.ExecuteRequest(userHREF, http.MethodGet,
+			types.MimeAdminUser, "error getting user: %s", nil, orgUser.User)
+
+		return orgUser, err
+	}
+	return nil, ErrorEntityNotFound
+}
+
+// FetchUserByName retrieves an user within an admin organization by name
+// Returns a valid user if it exists. If it doesn't, returns nil and ErrorEntityNotFound
+// If argument refresh is true, the AdminOrg will be refreshed before searching.
+// This is usually done after creating, modifying, or deleting users.
+// If it is false, it will search within the data already in memory (useful when
+// looping through the users and we know that no changes have occurred in the meantime)
+func (adminOrg *AdminOrg) FetchUserByName(name string, refresh bool) (*OrgUser, error) {
+	return adminOrg.fetchUserByNameOrId(name, true, true)
+}
+
+// FetchUserById retrieves an user within an admin organization by ID
+// Returns a valid user if it exists. If it doesn't, returns nil and ErrorEntityNotFound
+// If argument refresh is true, the AdminOrg will be refreshed before searching.
+// This is usually done after creating, modifying, or deleting users.
+// If it is false, it will search within the data already in memory (useful when
+// looping through the users and we know that no changes have occurred in the meantime)
+func (adminOrg *AdminOrg) FetchUserById(ID string, refresh bool) (*OrgUser, error) {
```

Comment:
Nit/naming convention as this lays a precedence for the new era of fetch functions (#211): `id string` instead of `ID string`

Reply:
done